From 4fc7411abfeef1d80b1ff97ef674322e7391e55d Mon Sep 17 00:00:00 2001
From: zgzxx <zhangguangzhi3@huawei.com>
Date: Wed, 10 Mar 2021 09:37:54 +0800
Subject: [PATCH 3/6] modify the prompt information

---
 tools/codegener/intel/Util.ml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/codegener/intel/Util.ml b/tools/codegener/intel/Util.ml
index 9629e54..f5f9afc 100644
--- a/tools/codegener/intel/Util.ml
+++ b/tools/codegener/intel/Util.ml
@@ -62,7 +62,7 @@ let dedup_list lst =
 let usage (progname: string) =
   eprintf "usage: %s [options] <file> ...\n" progname;
   eprintf "\n[options]\n\
---search-path <path>  Specify the search path of EDL files\n\
+--search-path <path>  Specify the search path of the dependent files of the EDL files\n\
 --use-prefix          Prefix untrusted proxy with Enclave name\n\
 --header-only         Only generate header files\n\
 --untrusted           Generate untrusted proxy and bridge\n\
-- 
2.27.0

